### PR TITLE
[WebGPU] https://webgpu.github.io/webgpu-samples/samples/renderBundles does not load

### PR DESCRIPTION
**Pull request by mwyrzykowski**
#### 0b7d33831395826f3fdba7417d104da2e50059b3
<pre>
[WebGPU] <a href="https://webgpu.github.io/webgpu-samples/samples/renderBundles">https://webgpu.github.io/webgpu-samples/samples/renderBundles</a> does not load
<a href="https://bugs.webkit.org/show_bug.cgi?id=261852">https://bugs.webkit.org/show_bug.cgi?id=261852</a>
&lt;radar://115810432&gt;

Reviewed by Tadeu Zagallo.

Fixup RenderBundle implementation to the point where
<a href="https://webgpu.github.io/webgpu-samples/samples/renderBundles">https://webgpu.github.io/webgpu-samples/samples/renderBundles</a>
works as expected.

* Source/WebGPU/WGSL/ConstantRewriter.cpp:
(WGSL::ConstantRewriter::materialize):
Add FIXME to workaround variant not containing the expected value.

* Source/WebGPU/WebGPU/Device.h:
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::Device::maxBuffersPlusVertexBuffersForVertexStage const):
(WebGPU::Device::vertexBufferIndexForBindGroup const):
Return the actual count of buffers.

* Source/WebGPU/WebGPU/RenderBundle.h:
(WebGPU::RenderBundle::create):
(WebGPU::RenderBundle::resources const):
* Source/WebGPU/WebGPU/RenderBundle.mm:
(-[ResourceUsageAndRenderStage initWithUsage:renderStages:]):
(WebGPU::RenderBundle::RenderBundle):
(WebGPU::RenderBundle::currentPipelineState const):
(WebGPU::RenderBundle::depthStencilState const):
(WebGPU::RenderBundle::cullMode const):
(WebGPU::RenderBundle::frontFace const):
(WebGPU::RenderBundle::depthClipMode const):
Save various state which is part of RenderBundle but not Metal&apos;s ICBs.

* Source/WebGPU/WebGPU/RenderBundleEncoder.h:
* Source/WebGPU/WebGPU/RenderBundleEncoder.mm:
(WebGPU::Device::createRenderBundleEncoder):
(WebGPU::RenderBundleEncoder::RenderBundleEncoder):
(WebGPU::RenderBundleEncoder::executePreDrawCommands):
(WebGPU::RenderBundleEncoder::draw):
(WebGPU::RenderBundleEncoder::drawIndexed):
(WebGPU::RenderBundleEncoder::drawIndexedIndirect):
(WebGPU::RenderBundleEncoder::drawIndirect):
(WebGPU::RenderBundleEncoder::finish):
(WebGPU::addResource):
(WebGPU::RenderBundleEncoder::setBindGroup):
(WebGPU::RenderBundleEncoder::setIndexBuffer):
(WebGPU::RenderBundleEncoder::setPipeline):
(WebGPU::RenderBundleEncoder::setVertexBuffer):
RenderBundleEncoder needs to restore previously set buffers on
earlier ICB commands.

* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::executeBundles):
Set state which is needed but not part of the ICBs.

Canonical link: <a href="https://commits.webkit.org/268454@main">https://commits.webkit.org/268454@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/890912c497c3cc0908482fa3e8f6a4d450496538

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19717 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20140 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/20757 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21612 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18430 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23400 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/20283 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/19999 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19937 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19939 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17146 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22466 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17116 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/24224 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18187 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18111 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22211 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18712 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/15856 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17868 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4723 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22221 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18537 "Built successfully") | | |

### CODE REVIEW

File: Source/WebGPU/WGSL/ConstantRewriter.cpp
Comment (mwyrzykowski) [inline Source/WebGPU/WGSL/ConstantRewriter.cpp:256]: @tadeuzagallo I filed an bug for this issue because working around the crash like this seems wrong

File: Source/WebGPU/WebGPU/RenderBundleEncoder.mm
Comment (mwyrzykowski) [inline Source/WebGPU/WebGPU/RenderBundleEncoder.mm:220]: We always have to set `m_icbDescriptor.maxVertexBufferBindCount` to the count from the device (which is usually 24) otherwise the generated shader buffer offsets will be incorrect

File: Source/WebGPU/WGSL/ConstantRewriter.cpp
Comment (djg) [reply Source/WebGPU/WGSL/ConstantRewriter.cpp:256]: With out diving in to work out where in the AST this is issue it coming from, I'm going to defer to @tadeuzagallo

Comment (tadeuzagallo) [reply Source/WebGPU/WGSL/ConstantRewriter.cpp:256]: Seems like we fail to compute a constant. It's fine with me to land as is and I can fix it after you land it.

**Review by tadeuzagallo — APPROVED** [review]

**Pull request merged**
